File: services/stream-processor/registry.go
package main

import (
	"sync"
	"time"
)

// seriesKey identifies one (device, metric) time series and carries the tag
// metadata InfluxDB needs — captured once when the series is first seen,
// since it doesn't change across samples for the same device+metric.
type seriesKey struct {
	FactoryID        string
	ProductionLineID string
	MachineID        string
	DeviceID         string
	SensorID         string
	Metric           string
}

func (k seriesKey) id() string { return k.DeviceID + "|" + k.Metric }

// seriesRegistry tracks one seriesBuffer per (device, metric) pair seen so
// far. Bounded by the number of distinct sensors in the system (1000 in this
// deployment) — it does not grow with message volume.
type seriesRegistry struct {
	mu        sync.Mutex
	buffers   map[string]*seriesBuffer
	keys      map[string]seriesKey
	maxWindow time.Duration
}

func newSeriesRegistry(maxWindow time.Duration) *seriesRegistry {
	return &seriesRegistry{
		buffers:   make(map[string]*seriesBuffer),
		keys:      make(map[string]seriesKey),
		maxWindow: maxWindow,
	}
}

func (r *seriesRegistry) record(key seriesKey, at time.Time, value float64) {
	id := key.id()

	r.mu.Lock()
	buf, ok := r.buffers[id]
	if !ok {
		buf = newSeriesBuffer(r.maxWindow)
		r.buffers[id] = buf
		r.keys[id] = key
	}
	r.mu.Unlock()

	buf.add(at, value)
}

// snapshot returns every tracked series' key and buffer, for a periodic
// flush to iterate without holding the registry lock during I/O.
func (r *seriesRegistry) snapshot() []struct {
	Key seriesKey
	Buf *seriesBuffer
} {
	r.mu.Lock()
	defer r.mu.Unlock()

	out := make([]struct {
		Key seriesKey
		Buf *seriesBuffer
	}, 0, len(r.buffers))
	for id, buf := range r.buffers {
		out = append(out, struct {
			Key seriesKey
			Buf *seriesBuffer
		}{Key: r.keys[id], Buf: buf})
	}
	return out
}

File: services/stream-processor/registry.py
"""Per-(device, metric) series tracking, the Python port of registry.go."""

from __future__ import annotations

import threading
from dataclasses import dataclass

from window import SeriesBuffer


@dataclass(frozen=True)
class SeriesKey:
    """Identifies one (device, metric) time series and carries the tag
    metadata InfluxDB needs — captured once when the series is first seen,
    since it doesn't change across samples for the same device+metric."""

    factory_id: str = ""
    production_line_id: str = ""
    machine_id: str = ""
    device_id: str = ""
    sensor_id: str = ""
    metric: str = ""

    def id(self) -> str:
        return f"{self.device_id}|{self.metric}"


@dataclass
class TrackedSeries:
    key: SeriesKey
    buf: SeriesBuffer


class SeriesRegistry:
    """Tracks one SeriesBuffer per (device, metric) pair seen so far.
    Bounded by the number of distinct sensors in the system (1000 in this
    deployment) — it does not grow with message volume."""

    def __init__(self, max_window_seconds: float):
        self._lock = threading.Lock()
        self._buffers: dict[str, SeriesBuffer] = {}
        self._keys: dict[str, SeriesKey] = {}
        self._max_window_seconds = max_window_seconds

    def record(self, key: SeriesKey, at, value: float) -> None:
        series_id = key.id()
        with self._lock:
            buf = self._buffers.get(series_id)
            if buf is None:
                buf = SeriesBuffer(self._max_window_seconds)
                self._buffers[series_id] = buf
                self._keys[series_id] = key
        buf.add(at, value)

    def snapshot(self) -> list[TrackedSeries]:
        """Returns every tracked series' key and buffer, for a periodic
        flush to iterate without holding the registry lock during I/O."""
        with self._lock:
            return [TrackedSeries(key=self._keys[sid], buf=buf) for sid, buf in self._buffers.items()]
